text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Dynamic Import Hints and Its Metadata
+status: ACTIVE
+
+created_at: 2020-03-16
+updated_at: 2020-03-16
+---
+
+# Hints Authoring Syntax and Its Metadata
+
+## Summary
+This RFC suggests authoring syntax for specifying dynamic import hints and describes its metadata shape.
+
+## Motivation
+Using hints allow component authors to declare the component preload context by parameterizing dynamic imports. The values from the hints can be used to optimize dependency resolution and pre-fetching.
+
+## Hint Invariants
+* hint values must be globally available and resolvable on the server side 
+* hint values must be statically verifiable
+* hint values must be finite and enumerable (ex: no Math.random())
+
+## Detailed design
+
+### Hints Syntax
+Hints are expressed as a comment inside a dynamic import expression. 
+The hint syntax is a block string with a *key *as a resource and a *value* as a resolved resource value
+
+```js
+/* <key>=<value> */
+```
+Hint comment ex:
+```js
+/* @salesforce/client/formFactor=Small */
+```
+
+Dynamic import with a hint ex:
+```js
+export async function example1() {
+    loadedModule = await import("lightning-foo"/* @salesforce/client/formFactor=Small */);
```

Comment:
There cases this RFC doesn't cover:
- What happen if the `import` first parameter is not a string literal but rather an identifier?
- Should the compilation fails if the comment doesn't respect the grammar?
- The proposed syntax only covers values being parsed a string? Do we want to allow in the future boolean, number or object values?
- Do we want to support multiple hints combinations? (AND / OR operators)
- What happen if the `key` or the `value` contains a string?

Reply:
From the team discussion:
1. If the first parameter is not a string literal, no metadata will be gathered. Neither for the dynamic import statement nor for any hint specified.
2. Compilation does not fail if the comment does not respect grammar. The hint will be ignored. An ESLint rule will be added as part of future work to flag incorrect hints.
3. The new syntax will be `"<hint key>": "<hint value>"` to disambiguate and accommodate string literals, boolean and number. 
   i. Object values will be considered for a future improvement.
4. Since the value part of the hint will be delimited with double quotes, it is left to the discretion of the component author to use operators.

In addition to these, the hint grammar will be restricted to use the following characters
1. hint key: `^"[a-zA-Z@][a-zA-Z0-9_-/]+"$`
  - Starts with an alphabet or @, followed by any number of alphanumeric characters, underscore, dash and forward slash
  - Begins and ends with a double quote 

2. delimiter: `[:]\s+`

3. hint value`: `^"[a-zA-Z0-9_-/@{}:]+"$`
 - Any alphanumeric character, underscore, dash, forward slash, @, curly braces and colon. We will not validate the hint value for its completeness or if its a valid json string etc.
 - Beings and ends with a double quote

Reply:
> This should probably be defined and standardized if we think it's useful.

IMO, the definition of the available hint key and values should be part of this RFC. On the implicit invariant of this RFC is that hints don't have any impact on the compilation output. Due to this, the RFC focussed on the grammar with hints so the compiler could parse and returned the hints as part of the metadata. How those hints are validated, post-processed and how they might impact the way modules are loaded is an application container concern.

But yes I agree that those hints should be standardized somewhere. Mobile team might give a first stab at it since they are the first consumer of this new feature. /cc @cseymourSF